File: scripts/R/fetch_collection_pmids.R
# ============================================================
# Fetch PMID list for dataset collection
# Purpose: Get PMIDs via manual list, PubMed collection scrape,
#          or Entrez esearch query (from config.yaml)
# Inputs: config.yaml, optional raw/pmid_list.csv (manual override)
# Outputs: raw/pmid_list.csv, raw/pubmed_manifest.csv
# ============================================================

suppressPackageStartupMessages({
  library(rentrez)
  library(readr)
  library(dplyr)
  library(tibble)
})

source("scripts/R/utils.R")

args <- commandArgs(trailingOnly = TRUE)
dataset_name <- if (length(args) >= 1) args[1] else "REGARDS"
repo_root <- if (length(args) >= 2) args[2] else "."

config <- load_dataset_config(dataset_name, repo_root)
dirs <- get_dataset_dirs(dataset_name, repo_root)
ensure_dataset_dirs(dirs)

pmid_path <- file.path(dirs$raw, "pmid_list.csv")
manifest_path <- file.path(dirs$raw, "pubmed_manifest.csv")

`%||%` <- function(x, y) if (is.null(x) || length(x) == 0) y else x

fetch_pubmed_collection_pmids <- function(collection_id, expected = NULL) {
  pmids <- character()
  page <- 1L
  base_url <- sprintf(
    "https://pubmed.ncbi.nlm.nih.gov/collections/%s/",
    collection_id
  )
  repeat {
    html <- system2(
      "curl",
      c(
        "-sL", "-A", "Mozilla/5.0", base_url,
        "-G", "-d", "sort=pubdate", "-d", paste0("page=", page)
      ),
      stdout = TRUE,
      stderr = FALSE
    )
    if (length(html) == 0) break
    html <- paste(html, collapse = "\n")
    page_pmids <- regmatches(
      html,
      gregexpr('data-article-id="([0-9]+)"', html, perl = TRUE)
    )[[1]]
    page_pmids <- gsub('data-article-id="([0-9]+)"', "\\1", page_pmids, perl = TRUE)
    if (length(page_pmids) == 0) break
    pmids <- c(pmids, page_pmids)
    cat("  page", page, ":", length(page_pmids), "PMIDs\n")
    page <- page + 1L
    if (!is.null(expected) && length(unique(pmids)) >= expected) break
    Sys.sleep(0.35)
  }
  unique(pmids)
}

if (file.exists(pmid_path) && file.info(pmid_path)$size > 50) {
  pmid_df <- read_csv(pmid_path, show_col_types = FALSE)
  pmids <- unique(as.character(pmid_df$pmid))
  source_type <- "manual"
  cat("Using existing pmid_list.csv:", length(pmids), "PMIDs\n")
} else {
  collection_id <- config$source$collection_id %||% NULL
  entrez_query <- config$source$automated_fetch_query %||%
    config$source$entrez_query %||% NULL
  expected <- config$source$expected_paper_count %||% NULL

  if (!is.null(collection_id) &&
      grepl("pubmed\\.ncbi\\.nlm\\.nih\\.gov/collections", config$source$collection_url %||% "")) {
    cat("Fetching PMIDs from PubMed collection", collection_id, "...\n")
    pmids <- fetch_pubmed_collection_pmids(collection_id, expected)
    source_type <- paste0("pubmed_collection_", collection_id)
    write_csv(tibble(pmid = pmids, source = source_type), pmid_path)
    cat("Wrote", length(pmids), "PMIDs to", pmid_path, "\n")
  } else if (!is.null(entrez_query)) {
    cat("Fetching PMIDs via Entrez:", entrez_query, "\n")
    search <- entrez_search(db = "pubmed", term = entrez_query, retmax = 10000)
    pmids <- search$ids
    source_type <- "esearch"
    write_csv(tibble(pmid = pmids, source = source_type), pmid_path)
    cat("Wrote", length(pmids), "PMIDs to", pmid_path, "\n")
  } else if (identical(dataset_name, "REGARDS")) {
    query <- '"NS041588"[Grant Number]'
    cat("Fetching PMIDs via Entrez (REGARDS default):", query, "\n")
    search <- entrez_search(db = "pubmed", term = query, retmax = 10000)
    pmids <- search$ids
    source_type <- "esearch"
    write_csv(tibble(pmid = pmids, source = source_type), pmid_path)
    cat("Wrote", length(pmids), "PMIDs to", pmid_path, "\n")
  } else {
    stop(
      "No PMID source configured for ", dataset_name,
      ". Set collection_id + collection_url, automated_fetch_query, or provide raw/pmid_list.csv"
    )
  }
}

fetch_time <- format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC")
manifest <- tibble(
  pmid = pmids,
  fetch_date = fetch_time,
  fetch_status = "pending",
  error_message = "",
  source = source_type
)
write_csv(manifest, manifest_path)

expected <- config$source$expected_paper_count
cat("Manifest:", nrow(manifest), "rows\n")
if (!is.null(expected) && length(pmids) != expected) {
  cat("NOTE: Retrieved", length(pmids), "PMIDs vs", expected,
      "in curated collection. Use manual pmid_list.csv for exact match.\n")
}

File: scripts/fetch_collection_pmids.py
#!/usr/bin/env python3
"""Fetch PMIDs for a My NCBI PubMed collection.

Priority:
1. Use datasets/{dataset}/raw/pmid_list.csv if present (manual export from collection)
2. Fall back to Entrez esearch using grant number query (automated, may differ from collection count)
"""

from __future__ import annotations

import argparse
import csv
import json
import re
import sys
import time
import urllib.parse
import urllib.request
from pathlib import Path


ESEARCH_URL = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi"

# REGARDS NIH grant — 894 papers vs 911 in curated collection (June 2026)
REGARDS_ESearch_QUERY = '"NS041588"[Grant Number]'


def fetch_esearch_pmids(query: str, retmax: int = 10000) -> list[str]:
    params = {
        "db": "pubmed",
        "term": query,
        "retmax": str(retmax),
        "retmode": "json",
    }
    url = f"{ESEARCH_URL}?{urllib.parse.urlencode(params)}"
    with urllib.request.urlopen(url, timeout=120) as resp:
        data = json.load(resp)
    result = data["esearchresult"]
    ids = result.get("idlist", [])
    count = int(result.get("count", 0))
    if count > len(ids):
        raise RuntimeError(f"eSearch returned {len(ids)} IDs but count={count}; increase retmax")
    return ids


def read_manual_pmids(path: Path) -> list[str]:
    text = path.read_text(encoding="utf-8")
    pmids = re.findall(r"\b(\d{7,8})\b", text)
    # preserve order, deduplicate
    seen: set[str] = set()
    ordered: list[str] = []
    for pmid in pmids:
        if pmid not in seen:
            seen.add(pmid)
            ordered.append(pmid)
    return ordered


def write_pmid_list(pmids: list[str], out_path: Path, source: str) -> None:
    out_path.parent.mkdir(parents=True, exist_ok=True)
    with out_path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=["pmid", "source"])
        writer.writeheader()
        for pmid in pmids:
            writer.writerow({"pmid": pmid, "source": source})


def main() -> int:
    parser = argparse.ArgumentParser(description="Fetch PMIDs for dataset collection")
    parser.add_argument("--dataset", required=True, help="Dataset name, e.g. REGARDS")
    parser.add_argument("--repo-root", default=".", help="Repository root")
    parser.add_argument(
        "--query",
        default=REGARDS_ESearch_QUERY,
        help="Entrez query when no manual pmid_list.csv",
    )
    args = parser.parse_args()

    repo = Path(args.repo_root)
    raw_dir = repo / "datasets" / args.dataset / "raw"
    manual_path = raw_dir / "pmid_list.csv"
    out_path = raw_dir / "pmid_list.csv"

    if manual_path.exists() and manual_path.stat().st_size > 50:
        # Already have manual list — don't overwrite unless forced
        pmids = read_manual_pmids(manual_path)
        source = "manual"
        print(f"Using existing {manual_path}: {len(pmids)} PMIDs")
    else:
        print(f"Fetching PMIDs via Entrez: {args.query}")
        time.sleep(0.4)
        pmids = fetch_esearch_pmids(args.query)
        source = "esearch"
        write_pmid_list(pmids, out_path, source)
        print(f"Wrote {len(pmids)} PMIDs to {out_path} (source: {source})")

    manifest_path = raw_dir / "pubmed_manifest.csv"
    fetch_time = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
    with manifest_path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(
            f, fieldnames=["pmid", "fetch_date", "fetch_status", "error_message", "source"]
        )
        writer.writeheader()
        for pmid in pmids:
            writer.writerow(
                {
                    "pmid": pmid,
                    "fetch_date": fetch_time,
                    "fetch_status": "pending",
                    "error_message": "",
                    "source": source,
                }
            )
    print(f"Wrote manifest: {manifest_path} ({len(pmids)} rows)")
    return 0


if __name__ == "__main__":
    sys.exit(main())
